subsys/bluetooth/audio/has.c

```diff
@@ -868,19 +856,97 @@ static int bt_has_cp_generic_update(struct has_client *client, uint8_t prev_inde
 	}
 }
 
+#if defined(CONFIG_BT_SETTINGS)
+struct client_context_store {
+	/* Last notified preset index */
+	uint8_t last_preset_index_known;
+};
+
+static int settings_set_cb(const char *name, size_t len_rd, settings_read_cb read_cb, void *cb_arg)
+{
+	struct client_context_store store;
+	struct client_context *context;
+	bt_addr_le_t addr;
+	ssize_t len;
+	int err;
+
+	if (!name) {
+		LOG_ERR("Insufficient number of arguments");
+		return -EINVAL;
+	}
+
+	err = bt_settings_decode_key(name, &addr);
+	if (err) {
+		LOG_ERR("Unable to decode address %s", name);
+		return -EINVAL;
+	}
+
+	context = context_find(&addr);
+	if (context == NULL) {
+		/* Find and initialize a free entry */
+		context = context_alloc(&addr);
+		if (context == NULL) {
+			LOG_ERR("Failed to allocate client_context for %s", bt_addr_le_str(&addr));
+			return -ENOMEM;
+		}
+	}
```

Comment:
So this part is to "restore" the client contexts based on the addresses in the `bt_settings`, right? 

This is effectively something we'd need to do for all services, as we can only use the `address` as the key for lookup, and the same approach used for bonding keys and `ccc_set` etc. 

But this also means that our storage looks something like 

```
bt/ccc/<addr><type>/val
bt/keys/<addr><type>/val
bt/has/<addr><type>/val
bt/XXX/<addr><type>/val
```

Which is also what `bt_settings_encode_key` mentions: 
```
/* Key format:
 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
 */
```

And maybe that's too late to change for the host, but otherwise it would have made much more sense to me to have done
```
bt/<addr><type>/<subsys>/<key>
```

The reason I'm bringing this up, is that we may want to do 
```
bt/audio/<addr><type>/<module>/<key>
```
like
```
bt/audio/<addr><type>/has/<key>
bt/audio/<addr><type>/csis/<key>
bt/audio/<addr><type>/tbs/<key>
```

Instead of 
```
bt/has/<addr><type>/<key>
bt/csis/<addr><type>/<key>
bt/tbs/<addr><type>/<key>
```

now when we have the chance, so that don't have duplicate `<addr><type>` entries for LE Audio at least. That will require a new `bt_audio_settings` API, but would also make it possible to omit the parsing of the `addr` for each module.

Reply:
What is the advantage of having
```
bt/audio/<addr><type>/has/<key>
bt/audio/<addr><type>/csis/<key>
bt/audio/<addr><type>/tbs/<key>
```
instead of 
```
bt/has/<addr><type>/<key>
bt/csis/<addr><type>/<key>
bt/tbs/<addr><type>/<key>
```
?

Reply:
> What is the advantage of having
> 
> ```
> bt/audio/<addr><type>/has/<key>
> bt/audio/<addr><type>/csis/<key>
> bt/audio/<addr><type>/tbs/<key>
> ```
> 
> instead of
> 
> ```
> bt/has/<addr><type>/<key>
> bt/csis/<addr><type>/<key>
> bt/tbs/<addr><type>/<key>
> ```
> 
> ?

I am actually unsure how it is stored in flash, by I was hoping that we would only need to store the `<addr><type>` once per device, rather than once per module (i.e. the address would only be stored once, instead of once per value), but maybe that is actually not the case. 

This would also be similar to how e.g. Windows (IIRC) and Linux store data per device:
```
ls -l /var/lib/bluetooth/
total 40
drwx------ 5 root root 4096 Jun 28 10:49 14:18:C3:7A:17:27
-rw------- 1 root root  133 Aug 16  2022 addresses
drwx------ 2 root root 4096 Sep 17  2021 F4:5A:12:BF:4F:2C
drwx------ 2 root root 4096 Sep 17  2021 FA:4C:4B:DB:D3:89
drwx------ 2 root root 4096 Aug 16  2022 static-D2:28:5D:1D:C2:72
drwx------ 2 root root 4096 May 31  2022 static-E0:BB:5A:36:CE:3E
drwx------ 2 root root 4096 May 31  2022 static-E3:5E:82:A4:F9:CC
drwx------ 2 root root 4096 May  8 15:41 static-EB:33:8C:03:0D:A2
drwx------ 2 root root 4096 Feb 16  2022 static-ED:45:67:44:84:4B
drwx------ 2 root root 4096 May  9 11:10 static-FD:60:4B:20:E3:C7
```

Reply:
> Typically, the service can call settings_load with the subtree name e.g. `bt/has` to read all the entries stored for this service. Some of the settings will have address as a part of the key, some not. With the solution you propose it would look like:
> 
> ```
> bt/audio/<addr><type>/has/<key>
> bt/audio/csis/<key>
> bt/audio/<addr><type>/tbs/<key>
> ```

That is true, it would complicate things for general (i.e. not device specific) data. 

If we keep it as is, then I think `bt/has` is probably better than `bt/audio/has`.